### PR TITLE
Add proxy configuration options

### PR DESCRIPTION
**Pull request by hferentschik**
See issue #65

### CODE REVIEW

**Comment by hferentschik** [comment]
@LalatenduMohanty support for proxies during registration ...

File: test/test_helper.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    true
+  end
+end
+
+module Registration
+  class FakeProvider
+    def initialize(*args)
+    end
+
+    def _initialize(*args)
```
Comment (budhrg) [inline test/test_helper.rb]: What is the use of this method? Seeing first time

File: features/registration.feature [hunk 2]
```diff
@@ -101,4 +101,4 @@ Feature: Booting VM with various registration settings
     Examples:
       | provider   |
       | virtualbox |
-      | libvirt    |
+  #    | libvirt    |
```
Comment (pvalena) [inline features/registration.feature]: Why is libvirt disabled?

File: Gemfile [hunk 3]
```diff
@@ -1,16 +1,26 @@
...
+
   gem 'cucumber', '~> 2.1'
   gem 'aruba', '~> 0.13'
   gem 'komenda', '~> 0.1.6'
   gem 'launchy'
   gem 'gem-compare'
   gem 'mechanize'
+
+  # build tool
+  gem 'rake', '10.4.2'
```
Comment (pvalena) [inline Gemfile]: Why do we need these specific version of rake and vagrant?

File: test/test_helper.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    end
+
+    def vm
+      VagrantPlugins::Kernel_V2::VMConfig.new
+    end
+  end
+end
+
+def fake_machine(options={})
+  env = options.fetch(:env, Vagrant::Environment.new({home_path: '/Users/hardy/work/developer-platform/vagrant/adb-vagrant-registration/build'}))
```
Comment (pvalena) [inline test/test_helper.rb]: The home_path is dependent on your system setup.

File: test/test_helper.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+require 'guests/redhat/cap/subscription_manager'
+require 'guests/redhat/cap/rhn_register'
+
+require 'minitest/autorun'
+require 'mocha/mini_test'
+
+require_relative 'support/fake_ui.rb'
+
+def fake_environment(options = { enabled: true })
+  { machine: fake_machine(options), ui: FakeUI, home_path: '/Users/hardy/work/developer-platform/vagrant/adb-vagrant-registration/build' }
```
Comment (pvalena) [inline test/test_helper.rb]: The home_path is dependent on your system setup.

**Comment by pvalena** [comment]
Is it necessary to have two testing frameworks?
Cannot these tests be implemented in Cucumber as well?

File: test/test_helper.rb (hunk 1 above)
Comment (hferentschik) [inline test/test_helper.rb]: I think we can remove it in this case. I'll try

File: features/registration.feature (hunk 2 above)
Comment (hferentschik) [inline features/registration.feature]: I commented it out temporarily and forgot about it. I enable it again since it can as well be controlled via the env variables. Good catch.

File: Gemfile (hunk 3 above)
Comment (hferentschik) [inline Gemfile]: There seems to be a bug in the latest version of Rake in combination with Minitest where test file loading will print a lot of warnings. Setting a fixed version for Rake will fix that.

File: test/test_helper.rb (hunk 4 above)
Comment (hferentschik) [inline test/test_helper.rb]: I'll change that. It might even not be necessary anymore. Initially this was part of the attempts to determine why the tests would not properly work, but it turned out to be mainly a Rake issue.

**Comment by hferentschik** [comment]
> Is it necessary to have two testing frameworks?

Yes

> Cannot these tests be implemented in Cucumber as well?

No. Cucumber is an acceptance test framework and works end to end, meaning I need to start the VM. It is also black box testing since it does not care about impl details of the plugin. Minitest on the other hand is a unit test framework to test single components w/o the need to start a VM. The more tests we can write in Minitest the better, since they run fast and can easily be used to debug the code.

File: test/test_helper.rb (hunk 1 above)
Comment (hferentschik) [inline test/test_helper.rb]: Actually, I am wrong. It is needed. A Vagrant provider defines it, so in order to "be" a fake Vagrant provider it needs to be there.

**Comment by hferentschik** [comment]
Pushed an update

**Comment by hferentschik** [comment]
Added some Cucumber tests as well

File: Gemfile (hunk 3 above)
Comment (pvalena) [inline Gemfile]: What about vagrant fixed in version 1.8.1, is it needed?

Comment (hferentschik) [inline Gemfile]: > What about vagrant fixed in version 1.8.1, is it needed? 

Yes. We could go to 1.8.4 (which is what we use in vagrant-service-manager atm), but referring to master does not work unless we decide to baseline development on Ruby 2.2. See also https://github.com/projectatomic/vagrant-service-manager/issues/268.

Bottom line, we need to pin the version.

Comment (LalatenduMohanty) [inline Gemfile]: @hferentschik lets go for the latest i.e. 1.8.4. I have not used 1.8.4 but there were some blocker issues in 1.8.1 and I am expecting some of them might have been fixed in 1.8.4.

File: features/registration.feature (hunk 2 above)
Comment (LalatenduMohanty) [inline features/registration.feature]: @hferentschik gentle reminder to un-comment libvirt.

File: README.md [hunk 6]
```diff
@@ -119,6 +120,27 @@ If you do not provide credentials, you will be prompted for them in the "up proc
 Please note the the interactive mode asks you for the preferred registration pair only
 of the configured manager.
 
+<a name="proxy-configuration"></a>
+### Proxy Configuration
+
+Proxies can be configured via the _proxy_, _proxyUser_ and _proxyPassword_ configuration options:
```
Comment (LalatenduMohanty) [inline README.md]: We should mention http proxy instead of just proxy.

File: README.md [hunk 7]
```diff
@@ -216,6 +238,15 @@ Note that the `auto_attach` option is set to false when using org/activationkey
   # Skip the registration (optional, skip if true, defaults to false)
   config.registration.skip
 
+  # Specify a HTTP proxy to use
+  config.registration.proxy
+
```
Comment (LalatenduMohanty) [inline README.md]: We need to add example with port in it.

File: test/vagrant-registration/cap/redhat/rhn_register_test.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    VagrantPlugins::GuestRedHat::Cap::RhnRegister.rhn_register_register(machine, FakeUI)
+
+    registration_command = machine.communicate.commands[:sudo][1]
+    assert_match(/--proxy='mongo:8080'/, registration_command, 'the proxy server and port should have been set')
+    assert_match(/--proxyUser='flash'/, registration_command, 'the proxy username have been set')
+    assert_match(/--proxyPassword='zarkov'/, registration_command, 'the proxy password should have been set')
+  end
+end
+
```
Comment (LalatenduMohanty) [inline test/vagrant-registration/cap/redhat/rhn_register_test.rb]: Is this file required for test or for actual code? looks like test but part of feature commit.

**Comment by LalatenduMohanty** [comment]
@hferentschik Looks  good to me except you need to address some review comments

File: README.md (hunk 6 above)
Comment (hferentschik) [inline README.md]: +1

File: Gemfile (hunk 3 above)
Comment (hferentschik) [inline Gemfile]: +1, makes sense to align

File: README.md (hunk 7 above)
Comment (hferentschik) [inline README.md]: ok

File: test/vagrant-registration/cap/redhat/rhn_register_test.rb (hunk 8 above)
Comment (hferentschik) [inline test/vagrant-registration/cap/redhat/rhn_register_test.rb]: good catch. This should go into test directory. Something went wrong here. I'll fix it.

Comment (hferentschik) [inline test/vagrant-registration/cap/redhat/rhn_register_test.rb]: Actually it is in the right directory. The UI is a bit misleading since it does not show the full path, but the file is in a _test_ sub-directory.

**Comment by hferentschik** [comment]
> Looks good to me except you need to address some review comments

Aligned with vagrant 1.8.4 and improved the README.

File: Gemfile (hunk 3 above)
Comment (voxik) [inline Gemfile]: Why not lock listen instead, if listen is the problem?

Comment (voxik) [inline Gemfile]: And TBH, locking version of Rake due to some warnings is a bit shortsighted because:

1. Its just warnings of test suite, so although they might be annoying, they makes no real harm
2. You will never discover that the warnings were fixed, since you locked the version and using the old version

Comment (hferentschik) [inline Gemfile]: > Why not lock listen instead, if listen is the problem?

That is not the big issue here. The big issue is that the dev branch of Vagrant switched to Ruby 2.2 as well. So even if you try to lock the listen version, you still need to deal with the change of Ruby runtime.

Comment (hferentschik) [inline Gemfile]: > And TBH, locking version of Rake due to some warnings is a bit shortsighted because:
> Its just warnings of test suite, so although they might be annoying, they makes no real harm

If I run the test I want a clean output. I don't want to see warnings. And in this case, even though the tests will eventually run, the warning seem to indicate that there is some recursive loading of files going on. I'd rather avoid that.

Also if I run the tests in my IDE it will fill up my console with warnings to an end that the IDE reports issues.

> You will never discover that the warnings were fixed, since you locked the version and using the old version

Rake is a development dependency. I use what makes sense. There is no need to "discover" updates. As long as the used version does what it should, all is good. Nothing stops you to once in a while try to test newer versions, but that can be done by a single developer. Different story with a runtime dependency, where you are more interested in getting the latest in order to get potential bug fixes and improvements. That said, even there I prefer to lock versions and do upgrades explicitly.

**Comment by strzibny** [comment]
@hferentschik Probably late to the party but why is `proxyUser` and others camelCased? If you want to follow the original naming, then it should be lowercase as the other options. These inconsistencies in API don't look good :/

**Comment by LalatenduMohanty** [comment]
@strzibny good point , I missed that during my review.

**Comment by strzibny** [comment]
@LalatenduMohanty I would suggest to fix this (and keep the fallback if there was a release already).
